src/dstack/_internal/core/backends/runpod/api_client.py

```diff
@@ -0,0 +1,280 @@
+from typing import Any, List, Optional
+
+import requests
+
+from dstack._internal.core.errors import BackendError, BackendInvalidCredentialsError
+
+API_URL = "https://api.runpod.io/graphql"
+
+
+class RunpodApiClient:
+    def __init__(self, api_key: str):
+        self.api_key = api_key
+
+    def validate_api_key(self) -> bool:
+        try:
+            self.get_user_details()
+        except BackendInvalidCredentialsError:
+            return False
+        return True
+
+    def get_user_details(self):
+        resp = self._make_request({"query": user_details_query, "variable": {}})
+        if resp.ok:
+            data = resp.json()
+            return data
+
+    def create_pod(
+        self,
+        name: str,
+        image_name: str,
+        gpu_type_id: str,
+        cloud_type: str = "ALL",
+        support_public_ip: bool = True,
+        start_ssh: bool = True,
+        data_center_id: Optional[str] = None,
+        country_code: Optional[str] = None,
+        gpu_count: int = 1,
+        volume_in_gb: int = 0,
+        container_disk_in_gb: Optional[int] = None,
+        min_vcpu_count: int = 1,
+        min_memory_in_gb: int = 1,
+        docker_args: str = "",
+        ports: Optional[str] = None,
+        volume_mount_path: str = "/runpod-volume",
+        env: Optional[dict] = None,
+        template_id: Optional[str] = None,
+        network_volume_id: Optional[str] = None,
+        allowed_cuda_versions: Optional[list] = None,
+    ) -> dict:
+        resp = self._make_request(
+            {
+                "query": generate_pod_deployment_mutation(
+                    name,
+                    image_name,
+                    gpu_type_id,
+                    cloud_type,
+                    support_public_ip,
+                    start_ssh,
+                    data_center_id,
+                    country_code,
+                    gpu_count,
+                    volume_in_gb,
+                    container_disk_in_gb,
+                    min_vcpu_count,
+                    min_memory_in_gb,
+                    docker_args,
+                    ports,
+                    volume_mount_path,
+                    env,
+                    template_id,
+                    network_volume_id,
+                    allowed_cuda_versions,
+                )
+            }
+        )
+        if resp.ok:
+            data = resp.json()
+            return data["data"]["podFindAndDeployOnDemand"]
+
+    def get_pod(self, pod_id: str):
```

Comment:
@Bihan, The return types should be annotated. The functions should be written so that they return on all code paths. Use `response.raise_for_status()` to raise errors. At the moment, unexpected errors are ignored, leading to other errors and making debugging hard.

Reply:
You do `response.raise_for_status()` in `_make_request()`, so it seems there is no need to check for `if resp.ok`.

Reply:
@r4victor Thank you for point out the issue. Yes `_make_request()` is the sole source as well as it handles the exceptions. Also I will annotate the return types and ensure functions return on all code paths. I will make changes and PR for the Fix.